### PR TITLE
KOGITO-8414 Added knative-service-discovery.adoc

### PR DESCRIPTION
**Pull request by hbelmiro**
**JIRA:** https://issues.redhat.com/browse/KOGITO-8414

This guide is for Knative custom function, which is implemented in https://github.com/kiegroup/kogito-runtimes/pull/2711

The example used by this guide is implemented in https://github.com/kiegroup/kogito-examples/pull/1526

**This PR should be merged only after the above PRs are merged.**

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributions doc](https://github.com/kiegroup/kogito-docs/blob/main/CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] The nav.adoc file has a link to this guide in the proper category
- [x] The index.adoc file has a card to this guide in the proper category, with a meaningful description

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+.Procedure
+
+. Add the `knative-serving` add-on dependency to your link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative/workflow[workflow] project.
++
+--
+
+[source,xml]
+----
+<dependency>
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:185]: please do not forget to update the existing SD doc.

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc [hunk 2]
```diff
@@ -0,0 +1,292 @@
+= Invoking Knative services from {context} custom functions
+:compat-mode!:
+// Metadata:
+:description: Explain what is and how the Knative service discovery works
+:keywords: kogito, workflow, quarkus, serverless, service-discovery, knative
+
+This document describes how to call Knative services using {context} custom functions. The procedure described in this document is based on the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[`serverless-workflow-custom-function-knative`] example application.
+
+The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```
Comment (kaldesai) [inline serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:9]:
```suggestion
The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+[NOTE]
+====
+`GET` requests are not yet supported.
+====
+
+[[about-namespaces]]
+=== About namespaces
+
+Note that in the above example you declared only the name of the service you wanted to call, but not a namespace. In this case, {product_name} will look for a Knative service in the same namespace the workflow service is running.
```
Comment (kaldesai) [inline serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:52]:
```suggestion
Note that in the above example, you declared only the name of the service you wanted to call, but not a namespace. In this case, {product_name} will look for a Knative service in the same namespace the workflow service is running.
```

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+  ]
+----
+
+In the above example, {product_name} will look for the `custom-function-knative-service` in the `my_different_namespace` namespace.
+
+=== About HTTP resource paths
+
+You can specify the HTTP resource path by adding the `path` property to the function's metadata.
+
+For instance, given the `/path/to/my/resource` HTTP resource path, you can define your function as:
```
Comment (kaldesai) [inline serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:73]:
```suggestion
For an instance, given the `/path/to/my/resource` HTTP resource path, you can define your function as follows:
```

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    }
+  ]
+----
+<1> The request payload is set in `arguments`.
+
+=== About the request timeout
+
+By default, the Knative service must respond within 10 seconds. You can use the `kogito.addon.knative-serving.request-timeout` property to configure this value.
+
+For instance, if you want to reduce the request timeout to 5 seconds, you must add the following to your `application.properties` file:
```
Comment (kaldesai) [inline serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:161]:
```suggestion
For an instance, if you want to reduce the request timeout to 5 seconds, you must add the following to your `application.properties` file:
```

**Review by kaldesai — APPROVED** [review]
@hbelmiro I have very minor suggestions. Your content looks good to me. Very well written! Thank you!

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc (hunk 1 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:185]: Done.

**Review by spolti — APPROVED** [review]

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc (hunk 4 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:73]: "For **_an_** instance"? Are you sure?
My intention is to say the same as "For example".

File: serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc (hunk 5 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:161]: Same as https://github.com/kiegroup/kogito-docs/pull/276#discussion_r1065875482

**Comment by hbelmiro** [comment]
@ricardozanini as we discussed offline, I've changed the structure to match camel-routes-integration.adoc and moved the guide to the "Integrations" menu.

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc [hunk 6]
```diff
@@ -0,0 +1,335 @@
+= Invoking Knative services from {context} custom functions
+:compat-mode!:
+// Metadata:
+:description: Explain what is and how the Knative service discovery works
+:keywords: kogito, workflow, quarkus, serverless, service-discovery, knative
+
+This document describes how to call Knative services using {context} custom functions. The procedure described in this document is based on the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[`serverless-workflow-custom-function-knative`] example application.
+
+The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:9]: Seems like you are repeating yourself here :)

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+
+[[enabling-knative-custom-functions]]
+== Enabing Knative {context} custom functions
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:12]:
```suggestion
== Enabling Knative {context} custom functions
```

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+
+You can enable Knative {context} custom functions in your project.
+
+.Prerequisites
+
+include::../cloud/common/_prerequisites.adoc[]
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:18]: Missing attributes:
```
{environment_prereq}

kubectl {kubectl_prereq}
```

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+.Procedure
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:20]: Display issue

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 8 above)
Comment (radtriste) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:18]: also do we really need all those prerequisites ?
In the procedure, you just mention to add the new dependency but you never show how to do it with Maven, quarkus CLI or kn CLI ...

**Review by radtriste — COMMENTED** [review]
The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO
and only the "Example" should be in this page

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+.Prerequisites
+
+include::../cloud/common/_prerequisites.adoc[]
+
+* You have the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative/custom-function-knative-service[custom-function-knative-service] project deployed on Knative. For more information on how to deploy a Quarkus project to Knative, see the https://quarkus.io/guides/deploying-to-kubernetes[Quarkus Kubernetes extension documentation].
+
+.Procedure
+
+. Add the `knative-serving` add-on dependency to your link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative/workflow[workflow] project.
```
Comment (MarianMacik) [inline serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:222]: The example could be mentioned earlier before the Prerequisites section, maybe. Something like we have in other docs, e.g. these steps are based on the serverless-workflow-custom-function-knative example.

**Comment by hbelmiro** [comment]
> The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO and only the "Example" should be in this page

The first part is explaining **Knative** custom functions (a specific type of custom function), while https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html is for custom functions in general. It explains how to create a custom function.

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 6 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc]: I removed the second one ;)

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 8 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc]: Yeah, we don't need them.
Removed.

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 9 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:14]: Fixed.

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 10 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:216]: This is mentioned in the guide's first paragraph. Do you think it's necessary to mention it again here?

**Comment by ricardozanini** [comment]
I agree with @radtriste, @hbelmiro. It would be nice to keep the ubiquity among the custom functions documentation.

**Comment by hbelmiro** [comment]
> > > The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO and only the "Example" should be in this page
> > 
> > 
> > @radtriste The first part is explaining **Knative** custom functions (a specific type of custom function), while https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html is for custom functions in general. It explains how to create a custom function.
> 
> well we are explaining there the Camel custom functions. could be good to centralize the custom functions type on that page. And then have a specific example in Integrations section, like @ricardozanini did for Camel

Done @radtriste.

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 11]
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
 
 === Function return values
 
-The Camel route is responsible to produce the return value in a way that the workflow can understand. The following are considered valid objects:
+The Camel route is responsible to produce the return value in a way that the workflow can understand.The following are considered valid objects:
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:300]:
```suggestion
The Camel route is responsible to produce the return value in a way that the workflow can understand. The following are considered valid objects:
```

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 12]
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
...
 
 include::../../pages/_common-content/camel-valid-responses.adoc[]
 
+[#knative-custom-function]
+== Knative custom function
+
+{product_name} provides an implementation of custom function through the `knative-serving` add-on to invoke Knative services.It allows you to have a static URI, defining a Knative service, which is used to perform HTTP requests.The Knative service defined in the URI is queried in the current Knative cluster and translated to a valid URL.
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:307]:
```suggestion
{product_name} provides an implementation of a custom function through the `knative-serving` add-on to invoke Knative services. It allows you to have a static URI, defining a Knative service, which is used to perform HTTP requests. The Knative service defined in the URI is queried in the current Knative cluster and translated to a valid URL.
```

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 13]
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
...
+
+[NOTE]
+====
+This section briefly exemplifies how to define and use Knative custom functions within your workflow application.For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:311]:
```suggestion
This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```

**Review by ricardozanini — APPROVED** [review]

File: serverlessworkflow/modules/ROOT/nav.adoc [hunk 14]
```diff
@@ -64,6 +64,7 @@
 ** xref:cloud/build-and-deploy-with-serverless-operator-on-kubernetes.adoc[Buiding and deploying a  {context} application on Kubernetes using the {product_name} Serverless Operator]
 * Integrations
 ** xref:integrations/camel-routes-integration.adoc[Integrating with Camel Routes]
+** xref:integrations/custom-functions-knative.adoc[Invoking Knative services from {context} custom functions]
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/nav.adoc:67]: I would say that you can remove `custom functions` from the title

**Review by radtriste — COMMENTED** [review]
I prefer much more this version :)

Just one minor comment about the page title. I think it can be reduced.

File: serverlessworkflow/modules/ROOT/nav.adoc (hunk 14 above)
Comment (hbelmiro) [reply serverlessworkflow/modules/ROOT/nav.adoc:67]: Done.

Comment (radtriste) [reply serverlessworkflow/modules/ROOT/nav.adoc:67]: please also update navbar and index :)

File: serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc (hunk 10 above)
Comment (MarianMacik) [reply serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:22]: @hbelmiro OK, I probably overlooked it. Thanks!

**Review by MarianMacik — APPROVED** [review]
Looks good once navbar and index are updated.

**Comment by hbelmiro** [comment]
> Looks good once navbar and index are updated.

Done.

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 15]
```diff
@@ -301,6 +301,175 @@ The Camel route is responsible to produce the return value in a way that the wor
...
 include::../../pages/_common-content/camel-valid-responses.adoc[]
 
+[#knative-custom-function]
+== Knative custom function
+
+{product_name} provides an implementation of a custom function through the `knative-serving` add-on to invoke Knative services. It allows you to have a static URI, defining a Knative service, which is used to perform HTTP requests. The Knative service defined in the URI is queried in the current Knative cluster and translated to a valid URL.
+
+[NOTE]
+====
+This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```
Comment (radtriste) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:311]:
```suggestion
This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow].
```

**Review by radtriste — APPROVED** [review]

**Pull request merged**